### DISCUSSION TITLE
Documenting gunicorn logging in Flask?

### DISCUSSION THREAD

**Discussion by pamelafox** [pallets/flask#5129] [Ideas]
I was wondering what you think of adding a section to the gunicorn docs page about logging. This seems to be the standard approach, and worked for me:
https://trstringer.com/logging-flask-gunicorn-the-manageable-way/

Specifically:

```
if __name__ != '__main__':
    gunicorn_logger = logging.getLogger('gunicorn.error')
    app.logger.handlers = gunicorn_logger.handlers
    app.logger.setLevel(gunicorn_logger.level)
```

It took some time to find that, however, and its not obvious from the gunicorn docs. I'm happy to draft a section if it seems like a good idea (and if that is indeed the best path).

**Comment by davidism** [comment]
I'm not really clear that it's a good idea to take all Gunicorn's logging handlers/levels and use them for Flask too. Typically you'd configure everything (Flask, Gunicorn, etc.) using `logging.config.dictConfig` ahead of time, not take Gunicorn's already-configured stuff and overwrite Flask's. That pattern is already shown in https://flask.palletsprojects.com/en/2.3.x/logging/, it's not Gunicorn specific.

So basically, that blog _kinda_ got it right, it shows configuring Flask loggers before creating the app. But it went off in the direction of copying other config rather than performing a full configuration, when it could have gone all in on configuring one root handler, for example

**Reply by pamelafox** [reply]
Okay, that's great to know. I'm trying to write samples that use Flask best practices, so I'll follow your suggestion, which works as well. The shortest version is probably:

```
logging.basicConfig(level=logging.DEBUG)
```

I'm putting that behind a guard so that it is only set in development to DEBUG level.

So I guess it doesn't warrant a section on the Gunicorn article, since it's no different than setup otherwise, but it's interesting how many articles/SO posts I found about the topic. Perhaps folks will now find this discussion, too :)

**Reply by AAAlvesJr** [reply]
@davidism, it seems the problem is that gunicorn has its own logger facility and configure that ahead of calling flask app. So it looks like we can't deal with gunicorn in the same way we do with other libraries. At least, it is not clear to me. Please, could you provide some example?

**Reply by davidism** [reply]
It's hard to provide an example because I don't know what problem you're describing. Gunicorn having logs is not a problem for Flask or any other code also having logs, because Python's `logging` is composable. I already linked to: https://flask.palletsprojects.com/en/stable/logging/, which talks about configuring logging in general with [`logging.config.dict_config`](https://docs.python.org/3/library/logging.config.html#logging.config.dictConfig).

**Reply by AAAlvesJr** [reply]
The goal is to get flask to push its messages to gunicorn's logger, as described in the page [https://trstringer.com/logging-flask-gunicorn-the-manageable-way/](https://trstringer.com/logging-flask-gunicorn-the-manageable-way/ ), cited by @pamelafox .
You said the way it is achieved there is not recommended. OK. But then, how to accomplish this using `logging.basicConf `?

I am asking because it is not obvious, because we get the gunicorn's logger already configured inside the flask app. 
Do you recommend making something like this
```
logging.basicConf(
...
handlers = logging.getLogger('gunicorn.error').handlers, 
level = logging.getLogger('gunicorn.error').level
...
)
```
???
